File: 42-trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    if (!height) {
        return 0;
    }

    let l = 0;
    let r = height.length - 1;
    let LeftMax = height[l];
    let RightMax = height[r];
    let res = 0;

    while (l < r) {
        if (LeftMax < RightMax) {
            l++;
            LeftMax = Math.max(LeftMax, height[l]);
            res += (LeftMax - height[l]);
        } else {
            r--;
            RightMax = Math.max(RightMax, height[r]);
            res += (RightMax - height[r]);
        }
    }
    return res;
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left = 0, right = height.length - 1;
        int res = 0;
        int leftMax = height[left];
        int rightMax = height[right];

        if (height == null || height.length == 0) {
            return 0;
        }

        while (left < right) {
            if (leftMax < rightMax) {
                left++;
                leftMax = Math.max(leftMax, height[left]);
                res += leftMax - height[left];
            } else {
                right--;
                rightMax = Math.max(rightMax, height[right]);
                res += rightMax - height[right];
            }
        }

        return res;
    }
}